report/table.go

```diff
@@ -13,54 +13,186 @@ import (
-	if len(labels) > MaxTableRows {
-		truncationCount := fmt.Sprintf("%d", len(labels)-MaxTableRows)
-		node = node.WithLatest(TruncationCountPrefix+prefix, mtime.Now(), truncationCount)
+	return node.WithTableTruncationInformation(prefix, len(rows))
+}
+
+// AddPrefixPropertyList appends arbitrary key-value pairs to the Node, returning a new node.
+func (node Node) AddPrefixPropertyList(prefix string, propertyList map[string]string) Node {
+	addedPropertiesCount := 0
+	for label, value := range propertyList {
+		if addedPropertiesCount >= MaxTableRows {
+			break
+		}
+		node = node.WithLatest(prefix+label, mtime.Now(), value)
+		addedPropertiesCount++
 	}
-	return node
+	return node.WithTableTruncationInformation(prefix, len(propertyList))
 }
 
-// ExtractTable returns the key-value pairs to build a table from this node
-func (node Node) ExtractTable(template TableTemplate) (rows map[string]string, truncationCount int) {
-	rows = map[string]string{}
-	truncationCount = 0
+// WithoutPrefix returns the string with trimmed prefix and a
+// boolean information of whether that prefix was really there.
+// NOTE: Consider moving this function to utilities.
+func WithoutPrefix(s string, prefix string) (string, bool) {
+	return strings.TrimPrefix(s, prefix), len(prefix) > 0 && strings.HasPrefix(s, prefix)
+}
+
+// ExtractMulticolumnTable returns the rows to build a multicolumn table from this node
+func (node Node) ExtractMulticolumnTable(template TableTemplate) (rows []Row) {
+	rowsByID := map[string]Row{}
+
+	// Itearate through the whole of our map to extract all the values with the key
+	// with the given prefix. Since multicolumn tables don't support fixed rows (yet),
+	// all the table values will be stored under the table prefix.
+	// NOTE: It would be nice to optimize this part by only iterating through the keys
+	// with the given prefix. If it is possible to traverse the keys in the Latest map
+	// in a sorted order, then having LowerBoundEntry(key) and UpperBoundEntry(key)
+	// methods should be enough to implement ForEachWithPrefix(prefix) straightforwardly.
 	node.Latest.ForEach(func(key string, _ time.Time, value string) {
-		if label, ok := template.FixedRows[key]; ok {
-			rows[label] = value
+		if keyWithoutPrefix, ok := WithoutPrefix(key, template.Prefix); ok {
+			ids := strings.Split(keyWithoutPrefix, TableEntryKeySeparator)
+			rowID, columnID := ids[0], ids[1]
+			// If the row with the given ID doesn't yet exist, we create an empty one.
+			if _, ok := rowsByID[rowID]; !ok {
+				rowsByID[rowID] = Row{
+					ID:      rowID,
+					Entries: map[string]string{},
+				}
+			}
+			// At this point, the row with that ID always exists, so we just update the value.
+			rowsByID[rowID].Entries[columnID] = value
 		}
-		if len(template.Prefix) > 0 && strings.HasPrefix(key, template.Prefix) {
-			label := key[len(template.Prefix):]
-			rows[label] = value
+	})
+
+	// Gather a sorted list of rows' IDs.
+	ids := make([]string, 0, len(rowsByID))
```

Comment:
In this particular case it seems easier to me to keep it as it is, since I don't know how to iterate through the hash with `for .. range` with index information. In general, `make([]string, 0, len(rowsByID))` seemed to be a more common pattern around the code I was changing, so I kept it like that everywhere for consistency, even though I'm not particularly biased towards it :)